odl/tomo/operators/xray_trafo.py

```diff
@@ -0,0 +1,285 @@
+        """
+        back, impl = self.backend.split('_')
+        if back == 'astra':
+            if impl == 'cpu':
+                return astra_cpu_forward_projector(x, self.geometry,
+                                                   self.range, out)
+            elif impl == 'cuda':
+                return astra_cuda_forward_projector(x, self.geometry,
+                                                    self.range, out)
+            else:
+                raise ValueError('unknown implementation {}.'.format(impl))
+        else:  # Should never happen
+            raise RuntimeError('backend support information is inconsistent.')
+
+    @property
+    def adjoint(self):
+        """Return the adjoint operator."""
+        return XrayBackProjector(self.domain, self.geometry, self.backend,
+                                 **self.kwargs)
+
+
+class XrayBackProjector(Operator):
+    """The adjoint of the discrete X-ray transform between `L^p` spaces."""
+
+    def __init__(self, discr_range, geometry, backend='astra', **kwargs):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        discr_range : `odl.DiscreteLp`
+            Reconstruction space, the range of the back-projector
+        geometry : `Geometry`
+            The geometry of the transform, contains information about
+            the operator domain. It needs to have a sampling grid for
+            motion and detector parameters.
+        backend : {'astra_cuda', 'astra_cpu'}, optional
+            Implementation back-end for the transform. Supported back-ends:
+            'astra_cpu': ASTRA toolbox using CPU, only 2D
+            'astra_cuda': ASTRA toolbox, using CUDA, 2D or 3D
+            Default: 'astra'
+        interp : {'nearest', 'linear'}
+            Interpolation type for the discretization of the operator range.
+            Default: 'nearest'
+        """
+        if not isinstance(discr_range, DiscreteLp):
+            raise TypeError('discretized range {!r} is not a `DiscreteLp`'
+                            ' instance.'.format(discr_range))
+
+        if not isinstance(geometry, Geometry):
+            raise TypeError('geometry {!r} is not a `Geometry` instance.'
+                            ''.format(geometry))
+
+        backend = str(backend).lower()
+        if backend not in _SUPPORTED_BACKENDS:
+            raise ValueError('backend {!r} not supported.'
+                             ''.format(backend))
+
+        if backend.startswith('astra'):
+            if not ASTRA_AVAILABLE:
+                raise ValueError('ASTRA backend not available.')
+            if not ASTRA_CUDA_AVAILABLE and backend == 'astra_cuda':
+                raise ValueError('ASTRA CUDA backend not available.')
+            if discr_range.dspace.dtype not in (np.float32, np.complex64):
+                raise ValueError('ASTRA support is limited to `float32` for '
+                                 'real and `complex64` for complex data.')
```

Comment:
I don't think so, and it would probably fail. We could easily support it, though, by doing the transform for real and imag separately. Or does ASTRA support complex single precision? I doubt it.